flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/filemerging/FileMergingSnapshotManagerBase.java

```diff
@@ -575,8 +576,51 @@ private void createManagedDirectory(Path managedPath) {
     @Override
     public void close() throws IOException {}
 
+    // ------------------------------------------------------------------------
+    //  restore
+    // ------------------------------------------------------------------------
+
+    @Override
+    public void restoreStateHandles(
+            long checkpointId, SubtaskKey subtaskKey, Stream<SegmentFileStateHandle> stateHandles) {
+
+        Set<LogicalFile> uploadedLogicalFiles;
+        synchronized (lock) {
+            uploadedLogicalFiles =
+                    uploadedStates.computeIfAbsent(checkpointId, id -> new HashSet<>());
+        }
+
+        stateHandles.forEach(
+                fileHandle -> {
+                    PhysicalFile physicalFile =
+                            new PhysicalFile(
+                                    null,
+                                    fileHandle.getFilePath(),
+                                    physicalFileDeleter,
+                                    fileHandle.getScope());
```

Comment:
Should we have shared `PhysicalFile` instance for multiple `LogicalFile`

Reply:
Yes, I have refactor this method. Please review FileMergingSnapshotManagerBase#restoreStateHandles.